rotkehlchen/exchanges/woo.py

```diff
@@ -0,0 +1,404 @@
+from rotkehlchen.assets.converters import asset_from_woo
+from rotkehlchen.constants import ZERO
+from rotkehlchen.errors.asset import UnknownAsset, UnsupportedAsset
+from rotkehlchen.errors.misc import RemoteError
+from rotkehlchen.errors.serialization import DeserializationError
+from rotkehlchen.exchanges.data_structures import AssetMovement, Trade
+from rotkehlchen.exchanges.exchange import ExchangeInterface, ExchangeQueryBalances
+from rotkehlchen.inquirer import Inquirer
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.serialization.deserialize import deserialize_asset_amount
+from rotkehlchen.types import ApiKey, ApiSecret, ExchangeAuthCredentials, Location, Timestamp
+from rotkehlchen.user_messages import MessagesAggregator
+from rotkehlchen.utils.misc import ts_now_in_ms
+from rotkehlchen.utils.mixins.cacheable import cache_response_timewise
+from rotkehlchen.utils.mixins.lockable import protect_with_lock
+from rotkehlchen.utils.serialization import jsonloads_dict
+
+if TYPE_CHECKING:
+    from rotkehlchen.db.dbhandler import DBHandler
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+API_KEY_ERROR_CODE_ACTION = {
+    '-1000': 'An unknown error occurred while processing the request.',
+    '-1001': 'The api key or secret is in wrong format.',
+    '-1002': 'The api key or secret is invalid.',
+}
+# Max limit for all API v2 endpoints
+API_MAX_LIMIT = 1000
+# user_transactions endpoint constants
+# Sort mode
+USER_TRANSACTION_SORTING_MODE = 'asc'
+# Starting `since_id`
+USER_TRANSACTION_MIN_SINCE_ID = 1
+# Trade type int
+USER_TRANSACTION_TRADE_TYPE = {2}
+# Asset movement type int: 0 - deposit, 1 - withdrawal
+USER_TRANSACTION_ASSET_MOVEMENT_TYPE = {0, 1}
+KNOWN_NON_ASSET_KEYS_FOR_MOVEMENTS = {
+    'datetime',
+    'id',
+    'type',
+    'fee',
+}
+
+
+class TradePairData(NamedTuple):
+    pair: str
+    base_asset_symbol: str
+    quote_asset_symbol: str
+    base_asset: AssetWithOracles
+    quote_asset: AssetWithOracles
+
+
+class Woo(ExchangeInterface):
+    """Woo exchange api docs:
+    https://docs.woo.org/#general-information
+
+    """
+    def __init__(
+            self,
+            name: str,
+            api_key: ApiKey,
+            secret: ApiSecret,
+            database: 'DBHandler',
+            msg_aggregator: MessagesAggregator,
+    ):
+        super().__init__(
+            name=name,
+            location=Location.WOO,
+            api_key=api_key,
+            secret=secret,
+            database=database,
+        )
+        self.base_uri = 'https://api.woo.org'
+        self.msg_aggregator = msg_aggregator
+        # NB: X-Auth-Signature, X-Auth-Nonce, X-Auth-Timestamp & Content-Type change per request
+        # X-Auth and X-Auth-Version are constant
+        self.session.headers.update({
+            'api_key': self.api_key,
+            'api_secret': self.secret,
+            'timestamp': str(ts_now_in_ms()),
```

Comment:
you don't need to set the timestamp here since when doing a query it will be updated